R/two.step.hybrid.R

```diff
@@ -1,170 +1,412 @@
-    
-    stopCluster(cl)
-
-    message("Recovery across batches")
-    
-    cl <- makeCluster(n.nodes)
-    registerDoParallel(cl)
-    #clusterEvalQ(cl, source("~/Desktop/Dropbox/1-work/apLCMS_code/new_proc_cdf.r"))
-    clusterEvalQ(cl, library(apLCMS))
-
-    #aligned<-foreach(batch.i=1:length(batches), .combine=cbind) %dopar%
-	for(batch.i in 1:length(batches))
-    {
-        this.fake<-batchwise[[batch.i]]$final.ftrs
-		this.fake.time<-batchwise[[batch.i]]$final.times
-        this.features<-batchwise[[batch.i]]$features2
-        this.medians<-apply(this.fake[,-1:-4],1,median)
-        
-        orig.time<-this.fake[,2]
-        adjusted.time<-fake2[[batch.i]][,2]
-        
-        this.pk.time<-this.aligned<-matrix(0, nrow=nrow(fake3$aligned.ftrs), ncol=ncol(this.fake)-4)
-        
-        # adjusting the time (already within batch adjusted)
-        
-        for(j in 1:length(this.features))
-        {
-            for(i in 1:nrow(this.features[[j]]))
-            {
-                diff.time<- abs(orig.time - this.features[[j]][i,2])
-                sel<-which(diff.time == min(diff.time))[1]
-                this.features[[j]][i,2] <- adjusted.time[sel]
-            }
+  }
+  return(within_batch$corrected_features)
+}
+
+compute_intensity_medians <- function(feature_table) {
+  stopifnot("sample_intensity" %in% colnames(feature_table))
+  feature_table <- dplyr::group_by(feature_table, mz, rt) %>%
+    dplyr::mutate(intensity = median(sample_intensity)) %>%
+    dplyr::ungroup()
+  return(feature_table)
+}
+
+bind_batch_label_column <- function(filenames, metadata) {
+  stopifnot(nrow(metadata) == length(filenames))
+
+  filenames <- as_tibble(filenames)
+  colnames(filenames)[1] <- "filename"
+  filenames <- mutate(filenames, sample_name = get_sample_name(filename))
+  filenames <- inner_join(filenames, metadata, by = "sample_name")
+  return(dplyr::select(filenames, -sample_name))
+}
+
+feature_recovery <- function(cluster,
+                             step_one_features,
+                             batchwise,
+                             filenames_batchwise,
+                             corrected,
+                             aligned,
+                             batches_idx,
+                             mz.tol,
+                             batch.align.mz.tol,
+                             batch.align.chr.tol,
+                             recover.mz.range,
+                             recover.chr.range,
+                             use.observed.range,
+                             min.bw,
+                             max.bw,
+                             recover.min.count) {
+  recovered <- tibble(
+    mz = numeric(),
+    rt = numeric(),
+    mz_min = numeric(),
+    mz_max = numeric()
+  )
+
+  for (batch_id in batches_idx)
+  {
+    this.fake <- step_one_features[[batch_id]]
+    this.fake.medians <- distinct(this.fake, mz, rt, intensity)$intensity
```

Comment:
Are these really the medians? If so, how do the distinct intensities form the medians?

Reply:
At this step `this.fake` is a long table where *intensity* is the column name of median intensities. Since each feature is repeated in multiple rows (a row for a sample) and values in *intensity* column is the same for all samples of a given feature, `distinct` selects a median intensity of each unique feature. 
In retrospect, calling a column that stores *median intensities* as *intensity* is confusing. Fixed that in 976e416.

Reply:
cool - how comes that the `distinct(...)` function actually computes the median intensities? I thought that it actually selects the unique values in the list?

Reply:
It doesn't compute the medians. The medians are already computed and stored in *intensity* column. `distinct(...)` only selects the median intensity of each unique feature.

Reply:
Refactored this part in 99ae4f5.